File: benches/competitive/bench.mjs
// Competitive perf harness: turbo-parsepdf (N-API) vs pdf-parse (pdf.js) on a
// fixture corpus. Build the addon first:
//   cargo build -p turbo-parsepdf-napi --release
//   node ../../crates/turbo-parsepdf-napi/scripts/copy-addon.mjs
// then: npm install && npm run bench

import { readFileSync, readdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const here = dirname(fileURLToPath(import.meta.url));
const corpus = join(here, "corpus");

async function adapters() {
  const a = {};
  try {
    const turbo = await import("../../crates/turbo-parsepdf-napi/index.js");
    a["turbo-parsepdf"] = (d) => turbo.parseToMarkdown(d);
  } catch {
    // addon not built; skip
  }
  try {
    const { default: pdfParse } = await import("pdf-parse");
    a["pdf-parse"] = async (d) => (await pdfParse(d)).text;
  } catch {
    // not installed; skip
  }
  return a;
}

async function best(fn, n = 10) {
  await fn();
  let ms = Infinity;
  for (let i = 0; i < n; i++) {
    const t = performance.now();
    await fn();
    ms = Math.min(ms, performance.now() - t);
  }
  return ms;
}

const libs = await adapters();
const names = Object.keys(libs);
const files = readdirSync(corpus).filter((f) => f.endsWith(".pdf")).sort();

console.log(`# turbo-parsepdf — competitive perf (best-of-N, ms)\n`);
console.log(`| file | ${names.join(" | ")} |`);
console.log(`|${"---|".repeat(names.length + 1)}`);
for (const f of files) {
  const data = readFileSync(join(corpus, f));
  const cells = [];
  for (const name of names) {
    try {
      cells.push((await best(() => libs[name](data))).toFixed(2));
    } catch (e) {
      cells.push(`ERR(${e.name})`);
    }
  }
  console.log(`| ${f} | ${cells.join(" | ")} |`);
}

File: benches/competitive-py/bench.py
#!/usr/bin/env python3
"""Competitive perf + accuracy harness: turbo-parsepdf vs the Python PDF stack.

Compares wall-clock text extraction (best-of-N) of `turbo_parsepdf` against
`pymupdf` (fitz), `pypdf`, and `pdfminer.six` on a fixture corpus, and checks
that turbo's text matches PyMuPDF (the reference) word-for-word.

Setup:
    python3 -m venv .venv && . .venv/bin/activate
    pip install pymupdf pypdf pdfminer.six
    # build + install turbo: cd ../../crates/turbo-parsepdf-py && maturin develop --release
Run:
    python bench.py [corpus_dir]   # default: ./corpus
"""

import io
import os
import sys
import time

CORPUS = sys.argv[1] if len(sys.argv) > 1 else os.path.join(os.path.dirname(__file__), "corpus")


def best(fn, n=10):
    fn()  # warm
    return min(_time(fn) for _ in range(n)) * 1000.0


def _time(fn):
    t = time.perf_counter()
    fn()
    return time.perf_counter() - t


def adapters():
    a = {}
    try:
        import turbo_parsepdf

        a["turbo-parsepdf"] = lambda d: turbo_parsepdf.parse_to_markdown(d)
    except ImportError:
        pass
    try:
        import fitz

        def mupdf(d):
            doc = fitz.open(stream=d, filetype="pdf")
            return "".join(p.get_text() for p in doc)

        a["pymupdf"] = mupdf
    except ImportError:
        pass
    try:
        import pypdf

        a["pypdf"] = lambda d: "".join(p.extract_text() for p in pypdf.PdfReader(io.BytesIO(d)).pages)
    except ImportError:
        pass
    try:
        from pdfminer.high_level import extract_text

        a["pdfminer.six"] = lambda d: extract_text(io.BytesIO(d))
    except ImportError:
        pass
    return a


def main():
    libs = adapters()
    files = sorted(f for f in os.listdir(CORPUS) if f.endswith(".pdf"))
    print(f"# turbo-parsepdf — competitive perf (best-of-N, ms)\n")
    header = "| file | " + " | ".join(libs) + " |"
    print(header)
    print("|" + "---|" * (len(libs) + 1))
    for f in files:
        data = open(os.path.join(CORPUS, f), "rb").read()
        cells = []
        for name, fn in libs.items():
            try:
                cells.append(f"{best(lambda: fn(data)):.2f}")
            except Exception as e:  # noqa: BLE001
                cells.append(f"ERR({type(e).__name__})")
        print(f"| {f} | " + " | ".join(cells) + " |")


if __name__ == "__main__":
    main()
